rbs/MethodTypeTranslator.cc

```diff
@@ -0,0 +1,307 @@
+#include "MethodTypeTranslator.h"
+#include "TypeTranslator.h"
+#include "absl/strings/escaping.h"
+#include "ast/AttrHelper.h"
+#include "ast/Helpers.h"
+#include "core/GlobalState.h"
+#include "core/errors/rewriter.h"
+
+using namespace sorbet::ast;
+
+namespace sorbet::rbs {
+
+namespace {
+
+core::NameRef expressionName(core::MutableContext ctx, const ast::ExpressionPtr *expr) {
+    core::NameRef name;
+
+    typecase(
+        *expr, [&](const ast::UnresolvedIdent &p) { name = p.name; },
+        [&](const ast::OptionalArg &p) { name = expressionName(ctx, &p.expr); },
+        [&](const ast::RestArg &p) { name = expressionName(ctx, &p.expr); },
+        [&](const ast::KeywordArg &p) { name = expressionName(ctx, &p.expr); },
+        [&](const ast::BlockArg &p) { name = expressionName(ctx, &p.expr); },
+        [&](const ast::ExpressionPtr &p) {
+            if (auto e = ctx.beginError(expr->loc(), core::errors::Rewriter::RBSError)) {
+                e.setHeader("Unexpected expression type: {}", p.showRaw(ctx));
+            }
+
+            name = ctx.state.enterNameUTF8("<error>");
+        });
```

Comment:
You may want to write this in a fashion similar to:

https://github.com/sorbet/sorbet/blob/1aecee9faa36281548916db8e47f36649105d21b/local_vars/local_vars.cc#L78-L122

It will probably be a good bit faster.